encoding/ccf/decode.go

```diff
@@ -0,0 +1,1879 @@
+/*
+ * Cadence - The resource-oriented smart contract programming language
+ *
+ * Copyright 2022-2023 Dapper Labs, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ccf
+
+import (
+	"fmt"
+	"math"
+	"math/big"
+	goRuntime "runtime"
+
+	"github.com/fxamacker/cbor/v2"
+	"github.com/onflow/cadence"
+	"github.com/onflow/cadence/runtime/common"
+	"github.com/onflow/cadence/runtime/errors"
+)
+
+// CBORDecMode
+//
+// See https://github.com/fxamacker/cbor:
+// "For best performance, reuse EncMode and DecMode after creating them."
+var CBORDecMode = func() cbor.DecMode {
+	decMode, err := cbor.DecOptions{
+		IntDec:           cbor.IntDecConvertNone,
+		MaxArrayElements: math.MaxInt64,
+		MaxMapPairs:      math.MaxInt64,
+		MaxNestedLevels:  math.MaxInt16,
+	}.DecMode()
+	if err != nil {
+		panic(err)
+	}
+	return decMode
+}()
+
+// A Decoder decodes CCF-encoded representations of Cadence values.
+type Decoder struct {
+	// CCF codec uses CBOR codec under the hood.
+	dec   *cbor.StreamDecoder
+	gauge common.MemoryGauge
+}
+
+// Decode returns a Cadence value decoded from its CCF-encoded representation.
+//
+// This function returns an error if the bytes represent CCF that is malformed,
+// invalid, or does not comply with requirements in the CCF specification.
+func Decode(gauge common.MemoryGauge, b []byte) (cadence.Value, error) {
```

Comment:
CCF decoder uses the same API as the decoder for JSON-Cadence Data Interchange:
- `Decode(common.MemoryGauge, []byte)`
- `NewDecoder(common.MemoryGauge, []byte)`
- `Decoder.Decode()`